## Fast CSV Queries

In this project, we will be practicing the time and space complexity analysis.  By taking both into account, we will utilize balanced amounts of pre-processing and efficient algorithm choice to speed up sorting/searching of csv data.


In this project, we will use an [existing laptop price dataset from Kaggle](https://www.kaggle.com/ionaskel/laptop-prices).  We will treat this data on laptops as if we were running a laptop store.

The columns contained in the data are as follows:
- ID: A unique identifier for the laptop.
- Company: The name of the company that produces the laptop.
- Product: The name of the laptop.
- TypeName: The type of laptop.
- Inches: The size of the screen in inches.
- ScreenResolution: The resolution of the screen.
- CPU: The laptop CPU.
- RAM: The amount of RAM in the laptop.
- Memory: The size of the hard drive.
- GPU: The graphics card name.
- OpSys: The name of the operating system.
- Weight: The laptop weight.
- Price: The price of the laptop.

In [2]:
# importing csv module
import csv

# reading dataset into jupyter
with open('laptops.csv') as file:
    data = list(csv.reader(file))
    
# printing first few rows
for row in data[:5]:
    print(row,'\n')

['Id', 'Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price'] 

['6571244', 'Apple', 'MacBook Pro', 'Ultrabook', '13.3', 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 2.3GHz', '8GB', '128GB SSD', 'Intel Iris Plus Graphics 640', 'macOS', '1.37kg', '1339'] 

['7287764', 'Apple', 'Macbook Air', 'Ultrabook', '13.3', '1440x900', 'Intel Core i5 1.8GHz', '8GB', '128GB Flash Storage', 'Intel HD Graphics 6000', 'macOS', '1.34kg', '898'] 

['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', '575'] 

['9722156', 'Apple', 'MacBook Pro', 'Ultrabook', '15.4', 'IPS Panel Retina Display 2880x1800', 'Intel Core i7 2.7GHz', '16GB', '512GB SSD', 'AMD Radeon Pro 455', 'macOS', '1.83kg', '2537'] 



In [9]:
# assigning header and rest of data to 'data'
header = data[0]
data = data[1:]

header

['Id',
 'Company',
 'Product',
 'TypeName',
 'Inches',
 'ScreenResolution',
 'Cpu',
 'Ram',
 'Memory',
 'Gpu',
 'OpSys',
 'Weight',
 'Price']

## Creating Inventory Class

The first thing we will do in this project is create a new class for inventory.  We will be updating our class definition as we need throughout the project.

In [10]:
# creating class inventory to represent our laptops in stock
class Inventory():
    def __init__(self, csv_filename):
        with open('laptops.csv') as file:
            data = list(csv.reader(file))
            self.header = data[0]
            self.rows = data[1:]
            
            # converting price column to integers
            for row in self.rows:
                price = row[-1]
                row[-1] = int(price)
                
                
# testing class
start_inv = Inventory('laptops.csv')

start_inv.header

['Id',
 'Company',
 'Product',
 'TypeName',
 'Inches',
 'ScreenResolution',
 'Cpu',
 'Ram',
 'Memory',
 'Gpu',
 'OpSys',
 'Weight',
 'Price']

In [11]:
for row in start_inv.rows[:5]:
    print(row,'\n')

['6571244', 'Apple', 'MacBook Pro', 'Ultrabook', '13.3', 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 2.3GHz', '8GB', '128GB SSD', 'Intel Iris Plus Graphics 640', 'macOS', '1.37kg', 1339] 

['7287764', 'Apple', 'Macbook Air', 'Ultrabook', '13.3', '1440x900', 'Intel Core i5 1.8GHz', '8GB', '128GB Flash Storage', 'Intel HD Graphics 6000', 'macOS', '1.34kg', 898] 

['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', 575] 

['9722156', 'Apple', 'MacBook Pro', 'Ultrabook', '15.4', 'IPS Panel Retina Display 2880x1800', 'Intel Core i7 2.7GHz', '16GB', '512GB SSD', 'AMD Radeon Pro 455', 'macOS', '1.83kg', 2537] 

['8550527', 'Apple', 'MacBook Pro', 'Ultrabook', '13.3', 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 3.1GHz', '8GB', '256GB SSD', 'Intel Iris Plus Graphics 650', 'macOS', '1.37kg', 1803] 



In [12]:
# length of dataset
len(start_inv.rows)

1303

One way that we may need to query this dataset for information is if a customer comes in with their laptop and we need an easy way to identify all of its different attributes.  

We will be adding a function to our class that can search our dataset by 'Id' called get_laptop_from_id().

In [13]:
import time 

class Inventory():
    def __init__(self, csv_filename):
        with open('laptops.csv') as file:
            data = list(csv.reader(file))
            self.header = data[0]
            self.rows = data[1:]
            
            # converting price column to integers
            for row in self.rows:
                price = row[-1]
                row[-1] = int(price)
    
    # function to search by laptop Id
    def get_laptop_from_id(self, laptop_id):
        for row in self.rows:
            if row[0] == laptop_id:
                return row
        return None
                
# testing class
start_inv = Inventory('laptops.csv')

# testing function get_laptop_from_id

# this id should exist 
print(start_inv.get_laptop_from_id('3362737'))

print('\n')

# this id should not exist
print(start_inv.get_laptop_from_id('3362736'))

['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', 575]


None


## Time Complexity and Algorithm Speedup

We can see that our function performs as it should.  The performance could be improved however.  The time complexity of this query is O(N), where N is the number of rows contained in the dataset.

By searching through a set, we can check in O(C) (constant time).  We can also achieve this with a dictionary as well, and since we need all the other laptop information return to us in the query, a dictionary is better suited for the task.

We only have 1300 laptops in this dataset, so this is largely irrelevant to this particular dataset, but it is good practice for when the length grows exponentially, and in todays world, big datasets are the future of data analytics.

We will start by storing each laptop Id as a key in a dictionary, with the entire row as it's value.  That way if we find the searched for "key" in the dictionary, it will return all information about the laptop.


In [14]:
class Inventory():
    def __init__(self, csv_filename):
        with open('laptops.csv') as file:
            data = list(csv.reader(file))
            self.header = data[0]
            self.rows = data[1:]
            
            # converting price column to integers
            for row in self.rows:
                price = row[-1]
                row[-1] = int(price)
            
            # creating dictionary with 'Id' as key and 'row' as value
            self.id_to_row = {}
            for row in self.rows:
                self.id_to_row[row[0]] = row
    
    # function to search by laptop Id
    def get_laptop_from_id(self, laptop_id):
        for row in self.rows:
            if row[0] == laptop_id:
                return row
        return None
    
    # function to search by laptop Id faster
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row:
            return row
        return None

# testing class
start_inv = Inventory('laptops.csv')
    
# testing modified search function

# this id should exist 
print(start_inv.get_laptop_from_id('3362737'))

print('\n')

# this id should not exist
print(start_inv.get_laptop_from_id('3362736'))

['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', 575]


None


## Comparing Speed

This new method should operate in constant time, and run much faster, at the expense of additional memory requirements to store the dictionary in RAM.  Using randomly generated id values from the random module, we will experiment to see how much faster the new search algorithm is at finding an id.

In [16]:
import numpy as np


# testing the time to search with the basic algorithm

# recording start time
start = time.time()

# generating list of random strings matching the length of the id in the dataset
a_list = [str(np.random.RandomState(42).randint(1000000, 9999999)) for _ in range(10000)]

# searching for ever key in the randomized list
for target in a_list:
    start_inv.get_laptop_from_id(target)

# recording end time
end = time.time()

total_time_basic = end-start
print('{} seconds total elapsed time'.format(total_time_basic))

0.84578537940979 seconds total elapsed time


In [17]:
# testing the time to search with the optimized algorithm

# recording start time
start = time.time()

# generating list of random strings matching the length of the id in the dataset
a_list = [str(np.random.RandomState(42).randint(1000000, 9999999)) for _ in range(10000)]

# searching for ever key in the randomized list
for target in a_list:
    start_inv.get_laptop_from_id_fast(target)

# recording end time
end = time.time()

total_time_fast = end-start
print('{} seconds total elapsed time'.format(total_time_fast))

0.04603767395019531 seconds total elapsed time


In [18]:
print("The optimized algorithm was approximately {}x as fast as the basic search"
      .format(round(total_time_basic / total_time_fast), 2))

The optimized algorithm was approximately 18x as fast as the basic search


In our make-believe store, let's assume we wanted to offer our customers a gift card good for up to $2500 dollars, and they could choose from three laptops:

- Laptop 1: $1339

- Laptop 2: $898

- Laptop 3: $575

The catch is that the customer can only make a single purchase with the gift card, so our store doesn't have to track remaining balances.  The customer might be upset that there is no combination of two laptops that equals $2500.

We will create a function that will check that given a gift card for "D" amount of dollars, there will be a way for them to spend the entire amount by purchasing up to two laptops.

We will be modifying our class to accept this new function check_promotion_dollars()

In [19]:
class Inventory():
    def __init__(self, csv_filename):
        with open('laptops.csv') as file:
            data = list(csv.reader(file))
            self.header = data[0]
            self.rows = data[1:]
            
            # converting price column to integers
            for row in self.rows:
                price = row[-1]
                row[-1] = int(price)
            
            # creating dictionary with 'Id' as key and 'row' as value
            self.id_to_row = {}
            for row in self.rows:
                self.id_to_row[row[0]] = row
    
    # function to search by laptop Id
    def get_laptop_from_id(self, laptop_id):
        for row in self.rows:
            if row[0] == laptop_id:
                return row
        return None
    
    # function to search by laptop Id faster
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    
    # function to check dollar amount for gift card purchase
    def check_promotion_dollars(self, dollars):
        for row in self.rows:
            price_1 = row[-1]
            if price_1 == dollars:
                return True
            for row in self.rows:
                if row[-1] + price_1 == dollars:
                    return True
        return False
    
    
# testing class
start_inv = Inventory('laptops.csv')

# testing function

# this amount should be found
print(start_inv.check_promotion_dollars(1000))

# this amount should not be found
print(start_inv.check_promotion_dollars(442))



True
False


## Time Complexity and Algorithm Speedup

Just as before with the "id search" function, we will optimize the check_promotion_dollars function by first creating a set with all prices contained in it.  This way, when we look for a laptop (and a second laptop), we can perform that search in constant time by allocating the set to memory.

In [20]:
class Inventory():
    def __init__(self, csv_filename):
        with open('laptops.csv') as file:
            data = list(csv.reader(file))
            self.header = data[0]
            self.rows = data[1:]
            
            # converting price column to integers
            for row in self.rows:
                price = row[-1]
                row[-1] = int(price)
            
            # creating dictionary with 'Id' as key and 'row' as value
            self.id_to_row = {}
            for row in self.rows:
                self.id_to_row[row[0]] = row
            
            # creating set to hold prices for check_promotion_dollars_fast functiom
            self.prices = set()
            for row in self.rows:
                self.prices.add(row[-1])
    
    # function to search by laptop Id
    def get_laptop_from_id(self, laptop_id):
        for row in self.rows:
            if row[0] == laptop_id:
                return row
        return None
    
    # function to search by laptop Id faster
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    
    # function to check dollar amount for gift card purchase
    def check_promotion_dollars(self, dollars):
        for row in self.rows:
            price_1 = row[-1]
            if price_1 == dollars:
                return True
            for row in self.rows:
                if row[-1] + price_1 == dollars:
                    return True
        return False
    
    # faster function to check dollar amount for gift card purchase
    def check_promotion_dollars_fast(self, dollars):
        if dollars in self.prices:
            return True
        for price in self.prices:
            if (dollars - price) in self.prices:
                return True
        
        return False

    
# testing class
start_inv = Inventory('laptops.csv')

# testing function

# this amount should be found
print(start_inv.check_promotion_dollars_fast(1000))

# this amount should not be found
print(start_inv.check_promotion_dollars_fast(442))


True
False


## Comparing Speed

This new method should operate in constant time, and run much faster, at the expense of additional memory requirements to store the set in RAM.  Using randomly generated id values from the random module, we will experiment to see how much faster the new check_promotion_dollars_fast method than the older, unoptimized method.

In [21]:
# testing the time to compute with the basic algorithm

# recording start time
start = time.time()

# generating list of random strings matching the length of the id in the dataset
a_list = [np.random.RandomState(42).randint(100, 5000) for _ in range(100)]

# searching for ever key in the randomized list
for target in a_list:
    start_inv.check_promotion_dollars(target)

# recording end time
end = time.time()

total_time_basic = end-start
print('{} seconds total elapsed time'.format(total_time_basic))

0.026015281677246094 seconds total elapsed time


In [22]:
# testing the time to search with the optimized algorithm

# recording start time
start = time.time()

# generating list of random strings matching the length of the id in the dataset
a_list = [np.random.RandomState(42).randint(100, 5000) for _ in range(100)]

# searching for ever key in the randomized list
for target in a_list:
    start_inv.check_promotion_dollars_fast(target)

# recording end time
end = time.time()

total_time_fast = end-start
print('{} seconds total elapsed time'.format(total_time_fast))

0.0009996891021728516 seconds total elapsed time


In [23]:
print("The optimized algorithm was approximately {}x as fast as the basic search"
      .format(round(total_time_basic / total_time_fast), 2))

The optimized algorithm was approximately 26x as fast as the basic search


## Finding A Laptop to Fit Any Budget

We would like to use a binary search to be able to list all laptops below a certain price "D" so a customer can find a laptop within their price range.  We know that binary search can only be performed if the data is first sorted by price.  Our function should speedily list all laptops within the customers price range.

We will implement this function by:

- sorting the price data using Pythons efficient sorted() function and sorting by the price column

- finding the first value great than dollar amount "D" using binary search

- returning all laptops with a value less than the previous step

In [10]:
class Inventory():
    def __init__(self, csv_filename):
        with open('laptops.csv') as file:
            data = list(csv.reader(file))
            self.header = data[0]
            self.rows = data[1:]
            
            # converting price column to integers
            for row in self.rows:
                price = row[-1]
                row[-1] = int(price)
            
            # creating dictionary with 'Id' as key and 'row' as value
            self.id_to_row = {}
            for row in self.rows:
                self.id_to_row[row[0]] = row
            
            # creating set to hold prices for check_promotion_dollars_fast functiom
            self.prices = set()
            for row in self.rows:
                self.prices.add(row[-1])
                
            # creating a sorted list forour find_first_laptop_more_expensive function
            self.rows_by_price = sorted(self.rows, key=(lambda row: row[-1]))
    
    # function to search by laptop Id
    def get_laptop_from_id(self, laptop_id):
        for row in self.rows:
            if row[0] == laptop_id:
                return row
        return None
    
    # function to search by laptop Id faster
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    
    # function to check dollar amount for gift card purchase
    def check_promotion_dollars(self, dollars):
        for row in self.rows:
            price_1 = row[-1]
            if price_1 == dollars:
                return True
            for row in self.rows:
                if row[-1] + price_1 == dollars:
                    return True
        return False
    
    # faster function to check dollar amount for gift card purchase
    def check_promotion_dollars_fast(self, dollars):
        if dollars in self.prices:
            return True
        for price in self.prices:
            if (dollars - price) in self.prices:
                return True
        
        return False
    
    # function to locate first laptop with higher price than target_price
    def find_first_laptop_more_expensive(self, target_price):
        range_start = 0                                   
        range_end = len(self.rows_by_price) - 1                       
        while range_start < range_end:
            range_middle = (range_end + range_start) // 2  
            price = self.rows_by_price[range_middle][-1]
            if price > target_price:
                range_end = range_middle
            else:                                          
                range_start = range_middle + 1
        if self.rows_by_price[range_start][-1] <= target_price:                  
        if price != target_price:                  
            return -1                                   
        return range_start

# testing class
start_inv = Inventory('laptops.csv')

# testing function

# this amount should be found
print(start_inv.find_first_laptop_more_expensive(1000))

# this amount should not be found
print(start_inv.find_first_laptop_more_expensive(442))


683
-1


So the first laptop greater than 1000 was at index 682.  Let's verify that the laptop at that index is greater than $1000 by printing out the few rows before and after it.

In [15]:
for row in start_inv.rows_by_price[681:685]:
    print(row, '\n')

['9505275', 'Dell', 'Inspiron 7579', '2 in 1 Convertible', '15.6', 'IPS Panel Full HD / Touchscreen 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'Windows 10', '2.191kg', 999] 

['6676297', 'HP', 'EliteBook 840', 'Notebook', '14', 'Full HD 1920x1080', 'Intel Core i5 6200U 2.3GHz', '4GB', '500GB HDD', 'Intel HD Graphics 520', 'Windows 10', '1.54kg', 1000] 

['8747948', 'Lenovo', 'ThinkPad T460', 'Notebook', '14', '1366x768', 'Intel Core i5 6200U 2.3GHz', '4GB', '508GB Hybrid', 'Intel HD Graphics 520', 'Windows 7', '1.70kg', 1002] 

['5550925', 'Dell', 'Latitude 5580', 'Notebook', '15.6', '1366x768', 'Intel Core i5 7300U 2.6GHz', '8GB', '500GB HDD', 'Intel HD Graphics 620', 'Windows 10', '1.9kg', 1008] 

